spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'stationary' , subtype:'eye-level' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .then(setupWebGLLayer)
+  .then(() => {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
+
+#### Position-disabled Subtype
+
+The origin of this subtype will be initialized at a position near the user's head at the time of creation.  `XRDevicePose` objects retrieved with this subtype will have varying orientation values but will always report `x`, `y`, `z` values to be `0`.
+
+```js
+let xrSession = null;
+let xrFrameOfReference = null;
+
+function onSessionStarted(session) {
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'stationary', subtype:'position-disabled' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .then(setupWebGLLayer)
+  .then(() => {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
+
+## Practical Usage Guidelines
+
+### Inline Sessions
+Inline sessions, by definition, do not require a user gesture or user permission to create, and as a result there must be strong limitations on the pose data that can be reported for privacy and security reasons.  Requests for an `XRBoundedFrameOfReference` or an `XRUnboundedFrameOfReference` will always be rejected on inline sessions.  Requests for `XRStationaryFrameOfReference` may succeed, but may also be rejected if the UA is unable provide any tracking information such as for an inline session on a desktop PC or a 2D browser window in a headset.
+
+Additionally, some inline experiences explicitly want tracking information disabled, such as a furniture viewer that will use pointer data to rotate the furniture.  Under both of these situations, there is no `XRFrameOfReference` needed.  Passing `null` as the `XRFrameOfReference` parameter to `getDevicePose()` or `getInputPose()` will return pose data whose values are based on the identity matrix.  Developers are then free to multiply in whatever transform they choose.
+
+### Ensuring hardware compatibility
+Immersive sessions will always be able to provide a `XRStationaryFrameOfReference`, but may not support other `XRFrameOfReference` types due to hardware limitations.  Developers are strongly encouraged to follow the spirit of [progressive enhanncement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement) and provide a reasonable fallback behavior if their desired `XRBoundedFrameOfReference` or `XRUnboundedFrameOfReference` is unavailable.  In many cases it will be adequate for this fallback to behave similarly to an inline preview experience.
+
+```js
+let xrSession = null;
+let xrFrameOfReference = null;
+
+function onSessionStarted(session) {
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'unbounded' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .catch(() => {
```

Comment:
Is the preferred method of feature detection here the rejection of a promise? I'd love to see one of either of the following. 

1. The ability to pass an array of objects into `xrSession.requestFrameOfRefrence` which would then return an `xrSession` Object with attributes indicating type and subtype.

or 

2. The ability to iterate over the FoR modes the UA supports and choose mode from there. I get that it's not ideal to give away information about the user's device prior to starting a session, but it seems like this info can be gained by doing the following in any case.

```js
function onSessionStarted (session) {
    return getFrameOfReference(session)
        .catch(() => {
            // some fallback
        })
        .then(frameOfReference => ({
            frameOfReference,
            session
        }))
        .then(setupWebGLLayer) // Some side effects here?
        .then(state => ({
            rid : session.requestAnimationFrame(onDrawFrame),
            ...state
        }));
}

async function getFrameOfReference (session, modes = [
    {type: "bounded"},
    {type: "unbounded"},
    {type: "standing", subtype: "eye-level"}
    // ...etc
]) {
    let frameOfReference;

    for (const mode of modes) {
        try {
            frameOfReference = await session.requestFrameOfReference(mode);
        } catch (e) {}
    }

    return frameOfReference ?
        Promise.resolve(frameOfReference) :
        Promise.reject(Error("no frame of reference available"));
}
```

It just seems cleaner from a dev perspective to be able to access which modes the user's device supports directly rather than having to handle a series of promises which may reject, and I think, in practice at least, most devs looking to build progressively enhanced XR experiences will end up doing something like the above.

Reply:
@TrevorFSmith had also suggest the "array of possible acceptable choices", it seems elegant.  May still need to deal with failure, but that's probably a UI that says "hey, sorry, you're out of luck" that that point.

Reply:
A method similar to `video` and `audio` element's [canPlayType](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#mime-types) method would be a good analogue for feature detection (in my mind at least). This is maybe a silly question, but are there any technical limitations what require feature detection to be asynchronous?

Reply:
Only tangentially-related to this question, but RE: promises, there is some concern that we can't use promises for many of the async APIs in WebXR due to very strict timing concerns with the resolution of the promise. Our current thinking is we will use callbacks instead. I am wondering if we should avoid using promises entirely throughout WebXR to avoid the inconsistency of sometimes using promises and sometimes using callbacks... @toji ?

Reply:
That sounds like a question for the TAG, but I suspect the answer is use promises when possible.

Reply:
I'm definitely in favor of using promises over callbacks for async functionality, and especially in favor of avoiding mixing the two. However, using promises for feature detection presents some awkward challenges to developers since the UA is going to throw in cases where is arguably shouldn't, namely feature detection. 

Let's consider the case where an application wants to support multiple FoR and serve content to users tailored to each. Ideally, the application would feature detect which FoR the user's device supports, then do one of two things. In the case where multiple FoR are supported (and where it would make sense), allow the user to select (or confirm) the FoR they want to use via an DOM UI. In the case where only one FoR is supported, simply transition into the XR view.

Given the current design, developers are forced into actually requesting each FoR, then catching an error if it isn't supported. It's certainly something that can be done easily enough, but it definitely feels a little clunky. The code I included above I think illustrates this in that there is an empty catch block, any time I see that it's a bit of a red flag for me. If the correct behavior in most cases for a catch statement is an empty block, then it really begs the question: why is the API throwing?

Reply:
@tencircles Yes, I've already incorporated Trevor's suggestion to request more than one FoR at at a time and you'll see that in the next push. https://github.com/immersive-web/webxr/pull/409#issuecomment-427435288

@lincolnfrog if you want to file a generic issue to discuss the approach WebXR is taking with promises vs. callbacks, please go ahead :)  However, let's not use this PR as the place to discuss it?

@tencircles Regarding the desire to feature detect... The reason this isn't supported had to do with avoiding fingerprinting concerns.  Knowing which FoR are supported is only something that should be queryable after the user has indicated intent to use the functionality, or in the case of immersive actually given permission. That said!  @toji and I are working on a subsequent PR to this one that may actually have a path to addressing this issue.  I'll be sure to get your feedback when we have it posted.

Reply:
After the IMCG call this morning, it sounds like I should update the language regarding floor-aligned FoR's to clearly state that the creation cannot cause UI elements to appear.